xarray/core/missing.py

```diff
@@ -20,44 +18,207 @@
-        cumulative_nans.diff(dim=dim, label="upper")
-        .reindex({dim: obj[dim]})
-        .where(valid)
-        .bfill(dim=dim)
-        .where(~valid, 0)
-        .fillna(index[-1] - valid_arange.max(dim=[dim]))
+    return _get_gap_right_edge(obj, dim, index, outside=True) - _get_gap_left_edge(
+        obj, dim, index, outside=True
     )
 
-    return nan_block_lengths
+
+def _get_max_gap_mask(
+    obj: T_Xarray, dim: Hashable, index: Variable, max_gap: int | float | np.number
+) -> T_Xarray:
+    nan_block_lengths = _get_nan_block_lengths(obj, dim, index)
+    return nan_block_lengths > max_gap
+
+
+def _get_gap_mask(
+    obj: T_Xarray,
+    dim: Hashable,
+    limit: T_GapLength | None = None,
+    limit_direction: LimitDirectionOptions = "both",
+    limit_area: LimitAreaOptions | None = None,
+    limit_use_coordinate=False,
+    max_gap: T_GapLength | None = None,
+    max_gap_use_coordinate=False,
+) -> T_Xarray | None:
+    # Input checking
+    ##Limit
+    if not is_scalar(limit):
+        raise ValueError("limit must be a scalar.")
+
+    if limit is None:
+        limit = np.inf
+    else:
+        if limit_use_coordinate is False:
+            if not isinstance(limit, Number | np.number):
+                raise TypeError(
+                    f"Expected integer or floating point limit since limit_use_coordinate=False. Received {type(limit).__name__}."
+                )
+        if _is_time_index(_get_raw_interp_index(obj, dim, limit_use_coordinate)):
+            limit = timedelta_to_numeric(limit)
+
+    ## Max_gap
+    if not is_scalar(max_gap):
+        raise ValueError("max_gap must be a scalar.")
+
+    if max_gap is None:
+        max_gap = np.inf
+    else:
+        if not max_gap_use_coordinate:
+            if not isinstance(max_gap, Number | np.number):
+                raise TypeError(
+                    f"Expected integer or floating point max_gap since use_coordinate=False. Received {type(max_gap).__name__}."
+                )
+
+        if _is_time_index(_get_raw_interp_index(obj, dim, max_gap_use_coordinate)):
+            max_gap = timedelta_to_numeric(max_gap)
+
+    # Which masks are really needed?
+    need_limit_mask = limit != np.inf or limit_direction != "both"
+    need_area_mask = limit_area is not None
+    need_max_gap_mask = max_gap != np.inf
+    # Calculate indexes
+    if need_limit_mask or need_area_mask:
+        index_limit = get_clean_interp_index(
+            obj, dim, use_coordinate=limit_use_coordinate
+        )
+        # index_limit = ones_like(obj) * index_limit
+    if need_max_gap_mask:
+        index_max_gap = get_clean_interp_index(
+            obj, dim, use_coordinate=max_gap_use_coordinate
+        )
+        # index_max_gap = ones_like(obj) * index_max_gap
+    if not (need_limit_mask or need_area_mask or need_max_gap_mask):
+        return None
+
+    # Calculate individual masks
+    masks = []
+    if need_limit_mask:
+        masks.append(
```

Comment:
I think it is difficult to tell mypy that limit cannot be a Hashable here do to the dynamic way of how it is handled here.
Maybe the best course of action is to add an assertion (not sure which exactly, maybe `assert not utils.is_hashable(limit)`, but maybe this is too restrictive?)
Easier might be to add a `#type: ignore[arg-type]` and a little comment as to why.